core/src/com/unciv/logic/civilization/diplomacy/DiplomacyManager.kt

```diff
@@ -187,6 +187,31 @@ class DiplomacyManager() : IsPartOfGameInfoSerialization {
     /** Total of each turn Science during Research Agreement */
     internal var totalOfScienceDuringRA = 0
 
+    // how quickly do we forget past relationships?
+    // used in AI denounce logic (values updated for all major civs, including humans, incase player resigns)
+    // a period of 1 is equivalent to only remembering our opinion from the last turn - very sudden drop in opinion is required to denounce
+    // a period of >1 means we keep some memory of opinions from previous turns
+    // if the old average is 10, new value is 25, and period is 3, then the new average will be 15, because the new value would contibute 1/3 to the new average
+    // (ema = exponential moving average)
+    @Readonly
+    private fun smoothedOpinionEmaPeriod() = 30f * civInfo.gameInfo.speed.modifier // value of 1 is useful for debugging
+    
+    // in old savefiles, if we have already met a civ, but the value has not been set, apply a very low initial value to prevent sudden mass denunciations
+    private val oldSavefileInitialOpinion = -250f
+    
+    // the smoothed opinion perpetually converges with the actual opinion, updated each turn
+    internal var smoothedOpinionOfOtherCiv = oldSavefileInitialOpinion
```

Comment:
Why can't this be entirely private? Also, maybe - lastSmoothedOpinionOfOtherCiv?